include/linux/simd_x86.h

```diff
@@ -0,0 +1,256 @@
+	[AVX]		= {1U,	0U,	1U << 28,	ECX	},
+	[AVX2]		= {7U,	0U,	1U << 5,	EBX	}
+};
+
+/*
+ * Check if CPU supports a feature
+ */
+static inline boolean_t
+_cpuid_check_feature(const cpuid_feature_desc_t *desc)
+{
+	unsigned int r[CPUID_REG_CNT];
+
+	if (__get_cpuid_max(0, NULL) >= desc->leaf) {
+		/*
+		 * __cpuid_count is needed to properly check
+		 * for AVX2. It is a macro, so return parameters
+		 * are passed by value.
+		 */
+		__cpuid_count(desc->leaf, desc->subleaf,
+			r[EAX], r[EBX], r[ECX], r[EDX]);
+		return (!!(r[desc->reg] & desc->flag));
+	}
+	return (B_FALSE);
+}
+
+#define	CPUID_FEATURE_CHECK(name, id) \
+static inline boolean_t	\
+cpuid_has_ ## name(void)\
+{	\
+	return (_cpuid_check_feature(&cpuid_features[id]));	\
+}
+
+/*
+ * Define functions for user-space CPUID features testing
+ */
+CPUID_FEATURE_CHECK(sse, SSE);
+CPUID_FEATURE_CHECK(sse2, SSE2);
+CPUID_FEATURE_CHECK(sse3, SSE3);
+CPUID_FEATURE_CHECK(ssse3, SSSE3);
+CPUID_FEATURE_CHECK(sse4_1, SSE4_1);
+CPUID_FEATURE_CHECK(sse4_2, SSE4_2);
+CPUID_FEATURE_CHECK(avx, AVX);
+CPUID_FEATURE_CHECK(avx2, AVX2);
+
+#endif /* !defined(_KERNEL) */
+
+/*
+ * Following functions should be called to determine whether CPU feature
+ * is supported. All functions are usable in kernel and user space.
+ * If an SIMD algorithm is using more than one instruction set
+ * all relevant feature test functions should be called.
+ */
+
+/*
+ * Check if SSE instruction set is available
+ */
+static inline boolean_t
+sse_available(void)
+{
+#if defined(_KERNEL)
+	return (!!boot_cpu_has(X86_FEATURE_XMM));
+#else
+	return (cpuid_has_sse());
+#endif
+}
+
+/*
+ * Check if SSE2 instruction set is available
+ */
+static inline boolean_t
+sse2_available(void)
+{
+#if defined(_KERNEL)
+	return (!!boot_cpu_has(X86_FEATURE_XMM2));
+#else
+	return (cpuid_has_sse2());
+#endif
+}
+
+/*
+ * Check if SSE3 instruction set is available
+ */
+static inline boolean_t
+sse3_available(void)
+{
+#if defined(_KERNEL)
+	return (!!boot_cpu_has(X86_FEATURE_XMM3));
+#else
+	return (cpuid_has_sse3());
+#endif
+}
+
+/*
+ * Check if SSSE3 instruction set is available
+ */
+static inline boolean_t
+ssse3_available(void)
+{
+#if defined(_KERNEL)
+	return (!!boot_cpu_has(X86_FEATURE_SSSE3));
+#else
+	return (cpuid_has_ssse3());
+#endif
+}
+
+/*
+ * Check if SSE4.1 instruction set is available
+ */
+static inline boolean_t
+sse4_1_available(void)
+{
+#if defined(_KERNEL)
+	return (!!boot_cpu_has(X86_FEATURE_XMM4_1));
+#else
+	return (cpuid_has_sse4_1());
+#endif
+}
+
+/*
+ * Check if SSE4.2 instruction set is available
+ */
+static inline boolean_t
+sse4_2_available(void)
+{
+#if defined(_KERNEL)
+	return (!!boot_cpu_has(X86_FEATURE_XMM4_2));
+#else
+	return (cpuid_has_sse4_2());
+#endif
+}
+
+/*
+ * Check if AVX instruction set is available
+ */
+static inline boolean_t
+avx_available(void)
+{
+#if defined(_KERNEL)
+	return (boot_cpu_has(X86_FEATURE_AVX) &&
+			boot_cpu_has(X86_FEATURE_OSXSAVE));
+#else
+	return (cpuid_has_avx());
```

Comment:
You still need to check osxsave for userspace, and you need to check if xsave saves avx registers using xgetbv.
https://github.com/tuxoko/zfs/commit/8baa4a5e0f43835a4233635d978f10fb7cbbb757#diff-0d2e10cd21fcf823e8e9a62a934e5115R48